klippa/src/lib.rs

```diff
@@ -7,67 +7,216 @@ mod parsing_util;
+            glyphset: IntSet::empty(),
+            num_h_metrics: 0,
+            num_output_glyphs: 0,
+            font_num_glyphs: 0,
+            unicode_to_new_gid_list: Vec::new(),
+            codepoint_to_glyph: HashMap::new(),
+        }
+    }
+
     pub fn new(input_gids: &IntSet<GlyphId>, input_unicodes: &IntSet<u32>, font: &FontRef) -> Self {
-        let gids = populate_unicodes_to_retain(input_gids, input_unicodes, font);
-        // remove invalid gids
-        let maxp = font.maxp().expect("Error reading maxp table");
-        let maxp = Maxp::from_table_ref(&maxp);
-        let gids: IntSet<GlyphId> = gids
-            .iter()
-            .filter(|gid| gid.to_u16() < maxp.num_glyphs)
-            .collect();
-        let num_glyphs = gids.len() as u16;
+        let mut this = Plan::empty();
+        // store font's num of glyphs in the plan
+        this.font_num_glyphs = get_font_num_glyphs(font);
+
+        this.populate_unicodes_to_retain(input_gids, input_unicodes, font);
+        this.populate_gids_to_retain(font);
+        this.num_output_glyphs = this.glyphset.len() as u16;
 
         // compute new h_metrics
         let hmtx = font.hmtx().expect("Error reading hmtx table");
         let hmtx = Hmtx::from_table_ref(&hmtx);
-        let new_h_metrics = compute_new_num_h_metrics(&hmtx, &gids);
+        this.num_h_metrics = compute_new_num_h_metrics(&hmtx, &this.glyphset);
 
-        Self {
-            glyph_ids: gids,
-            num_h_metrics: new_h_metrics,
-            num_output_glyphs: num_glyphs,
-        }
+        this
     }
-}
 
-fn populate_unicodes_to_retain(
-    input_gids: &IntSet<GlyphId>,
-    input_unicodes: &IntSet<u32>,
-    font: &FontRef,
-) -> IntSet<GlyphId> {
-    let mut gids = IntSet::empty();
-    gids.insert(GlyphId::new(0_u16));
-    gids.extend(input_gids.iter());
-
-    let cmap = font.cmap().expect("Error reading cmap table");
-    for cp in input_unicodes.iter() {
-        match cmap.map_codepoint(cp) {
-            Some(gid) => {
-                gids.insert(gid);
+    pub fn populate_unicodes_to_retain(
+        &mut self,
+        input_gids: &IntSet<GlyphId>,
+        input_unicodes: &IntSet<u32>,
+        font: &FontRef,
+    ) {
+        let charmap = font.charmap();
+        if input_gids.is_empty() && input_unicodes.len() < self.font_num_glyphs {
+            let cap = input_unicodes.len();
+            self.unicode_to_new_gid_list.reserve(cap);
+            self.codepoint_to_glyph.reserve(cap);
+            //TODO: add support for subset accelerator?
+
+            for cp in input_unicodes.iter() {
+                match charmap.map(cp) {
+                    Some(gid) => {
+                        self.codepoint_to_glyph.insert(cp, gid);
+                        self.unicode_to_new_gid_list.push((cp, gid));
+                    }
+                    None => {
+                        continue;
+                    }
+                }
```

Comment:
This is fine, but you could also write it as,

```suggestion
            for cp in input_unicodes.iter().flat_map(|cp| charmap.map(cp)) {
                self.codepoint_to_glyph.insert(cp, gid);
                self.unicode_to_new_gid_list.push((cp, gid));
```